src/main/java/net/neoforged/neoforge/common/extensions/IBlockExtension.java

```diff
@@ -215,11 +215,11 @@ default public boolean canHarvestBlock(BlockState state, BlockGetter level, Bloc
      * @param willHarvest True if Block.harvestBlock will be called after this, if the return in true.
      *                    Can be useful to delay the destruction of tile entities till after harvestBlock
      * @param fluid       The current fluid state at current position
-     * @return True if the block is actually destroyed.
+     * @return The new blockstate to pass to {@link Block#destroy} and {@link Block#playerDestroy} if this block was successfully replaced in world by {@link Level#setBlock}. Null if block was not replaced.
      */
-    default boolean onDestroyedByPlayer(BlockState state, Level level, BlockPos pos, Player player, boolean willHarvest, FluidState fluid) {
-        self().playerWillDestroy(level, pos, state, player);
-        return level.setBlock(pos, fluid.createLegacyBlock(), level.isClientSide ? 11 : 3);
+    default @Nullable BlockState onDestroyedByPlayer(BlockState state, Level level, BlockPos pos, Player player, boolean willHarvest, FluidState fluid) {
+        BlockState newState = self().playerWillDestroy(level, pos, state, player);
+        return level.setBlock(pos, fluid.createLegacyBlock(), level.isClientSide ? 11 : 3) ? newState : null;
```

Comment:
This feels *really* weird.  I think we might want to invent a new result container that conveys the intention of this method here, since we have two return values (the true/false "was the block broken" and the return from `playerWIllDestroy`)

edit: see above note on `canHarvestBlock` about what we want to do here.